### ISSUE TITLE
[HOLD for payment 2024-01-09] [$500] Theme - Skeleton placeholder in LHN does not switch theme when device theme is switched

### ISSUE THREAD

**Issue by izarutskaya** [Expensify/App#33163]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** v1.4.13-4
**Reproducible in staging?:** Y
**Reproducible in production?:** N
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Expensify/Expensify Issue URL:**
**Issue reported by:** Applause-Internal Team
**Slack conversation:** @

## Action Performed:
1. Log in to HT account (as it loads slower after login).
2. While the skeleton loading is showing, go to Settings > Preferences > Theme.
(If on Android or iOS, directly switch the dark mode settings from notification panel)
3. Switch to a different mode (if light, then dark; if dark, then light).
4. Close the RHP.

## Expected Result:
The skeleton placeholder in LHN will switch theme accordingly.

## Actual Result:
The skeleton placeholder in LHN does not switch theme.

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [x] Android: Native
- [x] Android: mWeb Chrome
- [x] iOS: Native
- [x] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [x] MacOS: Desktop

## Screenshots/Videos

https://github.com/Expensify/App/assets/115492554/ea777306-3ccf-41d8-9fb7-c5746aa86c09

![Bug6314470_1702642093031!Screenshot_20231215-200423_New_Expensify](https://github.com/Expensify/App/assets/115492554/28f2118e-cf0b-493e-bd3b-ed867d44a374)

</details>

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~016c5e43f31286b1d7</li>
        <li>Upwork Job ID: 1735650839183564800</li>
        <li>Last Price Increase: 2023-12-22</li>
<li>Automatic offers: </li>
<ul>
<li>hoangzinh | Reviewer | 28069280</li>
<li>ishpaul777 | Contributor | 28069281</li>
</ul></li>
    </ul>
</details>

**Comment by shubham1206agra** [comment]
## Proposal
### Please re-state the problem that we are trying to solve in this issue

Theme - Skeleton placeholder in LHN does not switch theme when device theme is switched

### What is the root cause of that problem?

This is because we are generating skeleton view items onLayout call in `OptionsListSkeletonView`. So this will not run again on changing theme.

### What changes do you think we should make in order to solve the problem?

1. Store component as a function that accepts theme and themestyles as argument, and use it as follows
```js
skeletonViewItems.push((theme, themeStyles) => (
                <SkeletonViewContentLoader
                    key={`skeletonViewItems${i}`}
                    animate={this.props.shouldAnimate}
                    height={CONST.LHN_SKELETON_VIEW_ITEM_HEIGHT}
                    backgroundColor={theme.skeletonLHNIn}
                    foregroundColor={theme.skeletonLHNOut}
                    style={themeStyles.mr5}
                >
                    <Circle
                        cx="40"
                        cy="32"
                        r="20"
                    />
                    <Rect
                        x="72"
                        y="18"
                        width="20%"
                        height="8"
                    />
                    <Rect
                        x="72"
                        y="38"
                        width={lineWidth}
                        height="8"
                    />
                </SkeletonViewContentLoader>
),
            );
```

2. And use it as map in View where state are loading
```js
<View>{this.state.skeletonViewItems.map((value) => value(this.props.theme, this.props.themeStyles))}</View>
```

### What alternative solutions did you explore? (Optional)

**Comment by flodnv** [comment]
@grgia can you have a look please?

**Comment by aswin-s** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Color of LHN skeleton view is not getting updated after changing theme.

### What is the root cause of that problem?

Skeleton view items are generated on 'onLayoutEvent' of 'OptionsSkeletonView'. And the OptionsSkeleton view is never unmounted even if we navigate away from the LHN view. Due to this the onLayout event never gets called again and the generated skeleton items retain the older theme color.

### What changes do you think we should make in order to solve the problem?

The most appropriate fix for this issue is to move the skeleton item rendering logic into the render method itself. Then the skeleton items would automatically respond to theme changes with out force rerendering. `OnLayout` event is only used to find the number of skeleton items that should be rendered. So only do the numItem calculation on layout event. Move the logic to render `SkeletonViewContentLoader` into render method and remove the `generateSkeletonViewItems` altogether.

Add an onLayout method that would simply set skeletonViewItems to an array of required length

```tsx
onLayoutChange(event) {
  const numItems = Math.ceil(event.nativeEvent.layout.height / CONST.LHN_SKELETON_VIEW_ITEM_HEIGHT);
  this.setState({skeletonViewItems: Array.from({length: numItems})});
}
```

Now update render method to render `SkeletonViewContentLoader` depending on the length of `skeletonViewItems`

```tsx

render() {
        return (
            <View
                style={[this.props.themeStyles.flex1, this.props.themeStyles.overflowHidden]}
                // update skeletonViewItems length on layout change
                onLayout={this.onLayoutChange.bind(this)}
            >
                <View>
                    {this.state.skeletonViewItems.map((_, i) => (
                        <SkeletonViewContentLoader
                            key={`skeletonViewItems${i}`}
                            animate={this.props.shouldAnimate}
                            height={CONST.LHN_SKELETON_VIEW_ITEM_HEIGHT}
                            backgroundColor={this.props.theme.skeletonLHNIn}
                            foregroundColor={this.props.theme.skeletonLHNOut}
                            style={this.props.themeStyles.mr5}
                        >
                            <Circle
                                cx="40"
                                cy="32"
                                r="20"
                            />
                            <Rect
                                x="72"
                                y="18"
                                width="20%"
                                height="8"
                            />
                            <Rect
                                x="72"
                                y="38"
                                width={i % 3 === 0 ? '100%' : i % 3 == 1 ? '50%' : '25%'}
                                height="8"
                            />
                        </SkeletonViewContentLoader>
                    ))}
                </View>
            </View>
        );
 }
```

**Result**

https://github.com/Expensify/App/assets/6880914/7e43c1b3-bb3c-4cf2-9603-7281fc5a813e

**Comment by Julesssss** [comment]
Not a deploy blocker, lets resolve at the regular priority

**Comment by sophiepintoraetz** [comment]
waiting on @grgia's take here.

**Comment by grgia** [comment]
@hoangzinh if you get an update, could you summarize in a comment here?

**Comment by sophiepintoraetz** [comment]
[Bumped in slack](https://expensify.slack.com/archives/C02NK2DQWUX/p1703275275231509)

**Comment by roryabraham** [comment]
> do you mind sharing the reason why you prefer to generate skeletonViewItems and store it in the component's state? In a first glance, I think it's because of optimization as @bernhardoj said above but I usually see the pattern where we generate a list/array of items (with a key) in component render function, leaving the rendering optimization part for React

As you've surmised, yes we store `skeletonViewItems` in state to prevent us from having to create new elements for each item on each render.

**Comment by roryabraham** [comment]
I don't really know if it's better to fix this by re-running `generateSkeletonViewItems` in `componentDidUpdate` or to just move the rendering logic for that into render, but since this component is still a class component and all it's dependencies are already using TS, here's what I think would be the best solution (written in TS):

```tsx
import React, {useMemo, useState} from 'react';
import {View} from 'react-native';
import {Circle, Rect} from 'react-native-svg';
import useTheme from '@hooks/useTheme';
import useThemeStyles from '@hooks/useThemeStyles';
import CONST from '@src/CONST';
import SkeletonViewContentLoader from './SkeletonViewContentLoader';

type OptionsListSkeletonViewProps = {
    shouldAnimate?: boolean;
};

function OptionsListSkeletonView({shouldAnimate = true}: OptionsListSkeletonViewProps) {
    const theme = useTheme();
    const themeStyles = useThemeStyles();

    const [numItems, setNumItems] = useState(0);
    const skeletonViewItems = useMemo(() => {
        const items = [];
        for (let i = 0; i < numItems; i++) {
            const step = i % 3;
            let lineWidth;
            switch (step) {
                case 0:
                    lineWidth = '100%';
                    break;
                case 1:
                    lineWidth = '50%';
                    break;
                default:
                    lineWidth = '25%';
            }
            items.push(
                <SkeletonViewContentLoader
                    key={`skeletonViewItems${i}`}
                    animate={shouldAnimate}
                    height={CONST.LHN_SKELETON_VIEW_ITEM_HEIGHT}
                    backgroundColor={theme.skeletonLHNIn}
                    foregroundColor={theme.skeletonLHNOut}
                    style={themeStyles.mr5}
                >
                    <Circle
                        cx="40"
                        cy="32"
                        r="20"
                    />
                    <Rect
                        x="72"
                        y="18"
                        width="20%"
                        height="8"
                    />
                    <Rect
                        x="72"
                        y="38"
                        width={lineWidth}
                        height="8"
                    />
                </SkeletonViewContentLoader>,
            );
        }
        return items;
    }, [numItems, shouldAnimate, theme, themeStyles]);

    return (
        <View
            style={[themeStyles.flex1, themeStyles.overflowHidden]}
            onLayout={(event) => {
                const newNumItems = Math.ceil(event.nativeEvent.layout.height / CONST.LHN_SKELETON_VIEW_ITEM_HEIGHT);
                if (newNumItems === numItems) {
                    return;
                }
                setNumItems(newNumItems);
            }}
        >
            <View>{skeletonViewItems}</View>
        </View>
    );
}

OptionsListSkeletonView.displayName = 'OptionsListSkeletonView';

export default OptionsListSkeletonView;
```

I believe this would solve the problem and simplify the component without sacrificing optimization. This way all of the `SkeletonViewContentLoader` items will only re-render if their dependencies change, and not simply because a parent component is re-rendering. See official docs on [memoizing individual JSX nodes](https://react.dev/reference/react/useMemo#memoizing-individual-jsx-nodes).

**Comment by shubham1206agra** [comment]
@hoangzinh Thanks for the review, but I think you should let @roryabraham handle the PR and you should review it rather than external contributor, since @roryabraham already gave the complete solution on which you agreed on.

**Comment by roryabraham** [comment]
Thanks everyone. I'm going to be pretty busy here, so I'm happy to let an external contributor handle the PR. I can also take over the review on behalf of @iwiznia and @grgia since I'll have good context on the solution.

**Comment by roryabraham** [comment]
PR merged

**Comment by sophiepintoraetz** [comment]
@hoangzinh - I need you to accept the offer so I can pay you! Thanks for being on it with the checklist.

**Comment by sophiepintoraetz** [comment]
All paid.
